src/installer/corehost/cli/bundle/info.cpp

```diff
@@ -0,0 +1,147 @@
+    trace::info(_X(".net core 3 compatibility mode: [%s]"), info.m_header.is_netcoreapp3_compat_mode() ? _X("Yes") : _X("No"));
+
+    the_app = &info;
+
+    return StatusCode::Success;
+}
+
+StatusCode info_t::process_header()
+{
+    try
+    {
+        const int8_t* addr = map_bundle();
+
+        reader_t reader(addr, m_bundle_size, m_header_offset);
+
+        m_header = header_t::read(reader);
+
+        unmap_bundle(addr);
+
+        return StatusCode::Success;
+    }
+    catch (StatusCode e)
+    {
+        return e;
+    }
+}
+
+void info_t::init_config(const pal::string_t& app_path)
+{
+    // Single-file bundles currently only support deps/runtime config json files
+    // named based on the app.dll. Any other name for these configuration files
+    // mentioned via the command line are assumed to be actual files on disk.
+    // 
+    // Supporting custom names for these config files is straightforward (with associated changes in bundler and SDK).
+    // There is no known use-case for it yet, and the facility is TBD.
+
+    m_base_path = get_directory(m_bundle_path);
+    pal::string_t deps_json_name = get_deps_from_app_binary(m_base_path, app_path);
+    pal::string_t runtimeconfig_json_name = get_runtime_config_path(m_base_path, get_filename_without_ext(app_path));
+
+    m_deps_json = config_t(deps_json_name, &m_header.deps_json_location());
+    m_runtimeconfig_json = config_t(runtimeconfig_json_name, &m_header.runtimeconfig_json_location());
+}
+
+int8_t* info_t::config_t::map(const pal::string_t& path, const location_t* &location)
+{
+    assert(is_single_file_bundle());
+
+    const bundle::info_t* app = bundle::info_t::the_app;
+    if (app->m_deps_json.matches(path))
+    {
+        location = app->m_deps_json.m_location;
+    }
+    else if (app->m_runtimeconfig_json.matches(path))
+    {
+        location = app->m_runtimeconfig_json.m_location;
+    }
+    else
+    {
+        return nullptr;
+    }
+
+    // When necessary to map the deps.json or runtimeconfig.json files, we map the whole single-file bundle,
+    // and return the address at the appropriate offset.
+    // This is because:
+    // * The host is the only code that is currently running, and attempting to map the bundle at this time
+    // * Files can only be memory mapped at page-aligned offsets, and in whole page units.
+    //   Therefore, mapping only portions of the bundle will involve align-down/round-up calculations, and associated offset adjustments.
+    //   We choose the simpler approach of rounding to the whole file
+    // * There is no performance limitation due to a larger sized mapping, since we actually only read the pages with relevant contents.
+    // * Files that are too large to be mapped (ex: that exhaust 32-bit virtual address space) are not supported. 
+
+    int8_t* addr = (int8_t*)pal::mmap_read(app->m_bundle_path);
```

Comment:
I know it's not part of this change yet... do we plan to "cache" the mapping? Once we start loading assemblies from memory I would expect to map the bundle once and hand out pointers from it as required...

Reply:
There's no plan to cache the mapping. 
* Closing the mapping enables renaming the app while it is running on Windows -- an idiom used to update apps in-place.
* The PE loader itself maps each section of the bundled assemblies at aligned addresses.

The only item that is cached is the bundle-manifest, so that bundle-probes from the runtime can be answered without need for re-processing the bundle headers.

Reply:
Good point - I forgot that runtime will create its own mapping anyway.